Fix test gap: mbedtls_cipher_setup_psa() with ECB

mprse opened the pull request:
## Description
Fixes https://github.com/ARMmbed/mbedtls/issues/4973.

Backport: [2.2x](https://github.com/ARMmbed/mbedtls/pull/5000)

This PR:
- adds test cases for  `AES-xxx-ECB PSA` (based on  `AES-xxx-CBC PSA`),
- adds missing code to support ECB mode for PSA. 

Test results after enabling `MBEDTLS_USE_PSA_CRYPTO` in `mbedtls_config.h (PSA part)`:
```
AES-128-CBC crypt Encrypt NIST KAT #1 PSA ......................... PASS
AES-128-CBC crypt Encrypt NIST KAT #2 PSA ......................... PASS
AES-128-CBC crypt Encrypt NIST KAT #3 PSA ......................... PASS
AES-128-CBC crypt Encrypt NIST KAT #4 PSA ......................... PASS
AES-128-CBC crypt Decrypt NIST KAT #1 PSA ......................... PASS
AES-128-CBC crypt Decrypt NIST KAT #2 PSA ......................... PASS
AES-128-CBC crypt Decrypt NIST KAT #3 PSA ......................... PASS
AES-128-CBC crypt Decrypt NIST KAT #4 PSA ......................... PASS
AES-192-CBC crypt Encrypt NIST KAT #1 PSA ......................... PASS
AES-192-CBC crypt Encrypt NIST KAT #2 PSA ......................... PASS
AES-192-CBC crypt Encrypt NIST KAT #3 PSA ......................... PASS
AES-192-CBC crypt Encrypt NIST KAT #4 PSA ......................... PASS
AES-192-CBC crypt Decrypt NIST KAT #1 PSA ......................... PASS
AES-192-CBC crypt Decrypt NIST KAT #2 PSA ......................... PASS
AES-192-CBC crypt Decrypt NIST KAT #3 PSA ......................... PASS
AES-192-CBC crypt Decrypt NIST KAT #4 PSA ......................... PASS
AES-256-CBC crypt Encrypt NIST KAT #1 PSA ......................... PASS
AES-256-CBC crypt Encrypt NIST KAT #2 PSA ......................... PASS
AES-256-CBC crypt Encrypt NIST KAT #3 PSA ......................... PASS
AES-256-CBC crypt Encrypt NIST KAT #4 PSA ......................... PASS
AES-256-CBC crypt Decrypt NIST KAT #1 PSA ......................... PASS
AES-256-CBC crypt Decrypt NIST KAT #2 PSA ......................... PASS
AES-256-CBC crypt Decrypt NIST KAT #3 PSA ......................... PASS
AES-256-CBC crypt Decrypt NIST KAT #4 PSA ......................... PASS
AES-128-ECB crypt Encrypt NIST KAT #1 PSA ......................... PASS
AES-128-ECB crypt Encrypt NIST KAT #2 PSA ......................... PASS
AES-128-ECB crypt Encrypt NIST KAT #3 PSA ......................... PASS
AES-128-ECB crypt Encrypt NIST KAT #4 PSA ......................... PASS
AES-128-ECB crypt Decrypt NIST KAT #1 PSA ......................... PASS
AES-128-ECB crypt Decrypt NIST KAT #2 PSA ......................... PASS
AES-128-ECB crypt Decrypt NIST KAT #3 PSA ......................... PASS
AES-128-ECB crypt Decrypt NIST KAT #4 PSA ......................... PASS
AES-192-ECB crypt Encrypt NIST KAT #1 PSA ......................... PASS
AES-192-ECB crypt Encrypt NIST KAT #2 PSA ......................... PASS
AES-192-ECB crypt Encrypt NIST KAT #3 PSA ......................... PASS
AES-192-ECB crypt Encrypt NIST KAT #4 PSA ......................... PASS
AES-192-ECB crypt Decrypt NIST KAT #1 PSA ......................... PASS
AES-192-ECB crypt Decrypt NIST KAT #2 PSA ......................... PASS
AES-192-ECB crypt Decrypt NIST KAT #3 PSA ......................... PASS
AES-192-ECB crypt Decrypt NIST KAT #4 PSA ......................... PASS
AES-256-ECB crypt Encrypt NIST KAT #1 PSA ......................... PASS
AES-256-ECB crypt Encrypt NIST KAT #2 PSA ......................... PASS
AES-256-ECB crypt Encrypt NIST KAT #3 PSA ......................... PASS
AES-256-ECB crypt Encrypt NIST KAT #4 PSA ......................... PASS
AES-256-ECB crypt Decrypt NIST KAT #1 PSA ......................... PASS
AES-256-ECB crypt Decrypt NIST KAT #2 PSA ......................... PASS
AES-256-ECB crypt Decrypt NIST KAT #3 PSA ......................... PASS
AES-256-ECB crypt Decrypt NIST KAT #4 PSA ......................... PASS
Cipher Corner Case behaviours ..................................... PASS

----------------------------------------------------------------------------
```

Open questions?

- Is ok that we got same results for CBC and ECB modes? 
- For ECB IV is irrelevant, but for CBC it is relevant, so why we are always testing the same vector ([all zeros](https://github.com/ARMmbed/mbedtls/blob/development/tests/suites/test_suite_cipher.aes.data#L1705-L1799))?


## Status
**READY**

## Migrations
NO

gilles-peskine-arm commented on tests/suites/test_suite_cipher.aes.data:
```diff
@@ -1798,6 +1798,102 @@ AES-256-CBC crypt Decrypt NIST KAT #4 PSA
 
+AES-128-ECB crypt Encrypt NIST KAT #1 PSA
+depends_on:MBEDTLS_USE_PSA_CRYPTO:MBEDTLS_AES_C
+test_vec_crypt:MBEDTLS_CIPHER_AES_128_ECB:MBEDTLS_ENCRYPT:"00000000000000000000000000000000":"00000000000000000000000000000000":"80000000000000000000000000000000":"3ad78e726c1ec02b7ebfe92b23d9ec34":0:1
```
The IV argument is unused for ECB. Please use an empty string so as to reduce the clutter in the test data and not give the impression that the IV is relevant (that's what we did for the non-PSA ECB test cases that use `test_vec_crypt`).

AndrzejKurek commented:
I think that this PR might need a changelog entry, since it changes default behaviour for a certain cipher mode.

mstarzyk-mobica reviewed (dismissed):
Looks alright

gilles-peskine-arm reviewed (changes requested):
> Is ok that we got same results for CBC and ECB modes?

The first output block of AES-CBC encryption is AES(IV xor P1) where P1 is the first plaintext block. For AES-ECB it's AES(P1). This is the same value as CBC with IV=all-bits-zero.

> For ECB IV is irrelevant, but for CBC it is relevant, so why we are always testing the same vector (all zeros)?

That's a very longstanding defect in our test data. Filed as https://github.com/ARMmbed/mbedtls/issues/4999

mprse commented on tests/suites/test_suite_cipher.function:
```diff
@@ -950,6 +950,14 @@ void test_vec_crypt( int cipher_id, int operation, data_t *key,
     {
         PSA_ASSERT( psa_crypto_init( ) );
+
+        if (cipher_id == MBEDTLS_CIPHER_AES_192_ECB ||
```
I'm a bit confused now. I understood that ECB should support only 128 bit size (and return error otherwise):

> Note that for ECB, unlike other modes and unlike psa_cipher_xxx, mbedtls_cipher_crypt for ECB only accepts 128-bit inputs. So for inputs of different lengths, we expect an error (but we still want to have a test, to ensure that we aren't getting garbage output or a crash).

gilles-peskine-arm replied:
That's 128-bit *inputs*. The “ECB” functions in the cipher module actually only encrypt a single block, and the size of a block is always 128 bits for AES. The number in `MBEDTLS_CIPHER_AES_192_ECB` refers to the size of the *key*, and all three possible sizes for AES (128, 192 or 256 bits) are supported.

mprse commented:
Addressed review comments.
Confirmed that tests still pass for other key sizes.

@AndrzejKurek I checked the `ChangeLog` file but I'm not sure what and where should I add in this file. Generally this PR just adds tests and provides fix for PSA ECB mode.

gilles-peskine-arm reviewed (commented):
Looks good to me so far. Since this is fixing a bug, I think it needs a changelog entry.

To add a changelog entry, create a file in `ChangeLog.d`. See `ChangeLog.d/00README.md` for more details. Suggested content (under Bugfix):

> Fix mbedtls_cipher_crypt for AES-ECB when MBEDTLS_USE_PSA_CRYPTO is enabled.

Also this needs to be backported to the 2.2x branch. Please make a pull request with similar commits to the `development_2.2x` branch.

mprse commented:
```
FAILED (605 / 606 tests (184 skipped))
^^^^test_tls13_experimental_with_ecp_restartable: ssl-opt.sh (TLS 1.3 experimental): "$@" -> 1^^^^
```

It has passed before I committed the change-log, but I doubt that this is the reason. Maybe I need to rebase PR?

gilles-peskine-arm commented:
There's an unusually high load on the CI today and it's causing a lot of timeouts.

Generally speaking, the intermittent failure of "DTLS client reconnect from same port: reference" under load is a [known bug in our test scripts](https://github.com/ARMmbed/mbedtls/issues/3377).

gilles-peskine-arm commented:
Actually I was looking at the wrong log. This isn't #3377.
```
14:18:22 Session resume using cache, DTLS: openssl client ....................... RETRY RETRY FAIL
14:18:22   ! bad client exit code (expected 0, got 143)
14:18:22   ! outputs saved to o-XXX-149.log
```
However it does look like a timeout caused by heavy load.

gilles-peskine-arm commented:
After yesterday's spike, there are still a lot of failures on the CI due to timeouts and unrelated to the content of the PR. Between https://jenkins-mbedtls.oss.arm.com/job/mbed-tls-pr-merge/job/PR-4996-merge/10, https://jenkins-mbedtls.oss.arm.com/job/mbed-tls-pr-merge/job/PR-4996-merge/8 and https://jenkins-mbedtls.oss.arm.com/job/mbed-tls-pr-head/job/PR-4996-head/6, I'm satisfied that there's enough coverage to claim that this PR is passing, so I'm going to go ahead and merge.

mprse commented:
I tried to reproduce the fail locally today. Locally test passes. Logs below: 
[log.txt](https://github.com/ARMmbed/mbedtls/files/7267378/log.txt)
